File: Python/200. Number of Islands.py
grid = [
    ["1", "1", "0", "0", "0"],
    ["1", "1", "0", "0", "0"],
    ["0", "0", "1", "0", "0"],
    ["0", "0", "0", "1", "1"]
]
w = len(grid[0])
h = len(grid)
steps = [[0, 1], [1, 0], [0, -1], [-1, 0]]
n = 0


def checkIsland(i, j):
    if grid[i][j] == "1":
        grid[i][j] = "0"
        for step in steps:
            if (grid[max(0, min(h - 1, i + step[0]))][max(0, min(w - 1, j + step[1]))] == "1"):
                checkIsland(
                    max(0, min(h - 1, i + step[0])), max(0, min(w - 1, j + step[1])))


for i in range(h):
    for j in range(w):
        if (grid[i][j] == "1"):
            checkIsland(i, j)
            n += 1
print(n)

File: Javascript/200. Number of Islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function (grid) {
  let res = 0;
  const width = grid.length;
  const height = grid[0].length;

  const safeIndex = (val, max) => Math.min(Math.max(val, 0), max);

  const traverse = (i, j) => {
    const step = [
      [0, 1],
      [1, 0],
      [-1, 0],
      [0, -1],
    ];

    step.forEach(([x, y]) => {
      const [newX, newY] = [
        safeIndex(i + x, width - 1),
        safeIndex(j + y, height - 1),
      ];
      if (grid[newX][newY] == "1") {
        grid[newX][newY] = "0";
        traverse(newX, newY);
      }
    });
  };

  for (let i = 0; i < width; i++) {
    for (let j = 0; j < height; j++) {
      if (grid[i][j] == "1") {
        grid[i][j] = "0";
        traverse(i, j);
        res += 1;
      }
    }
  }

  return res;
};
